Issue #24577: edit commonly-used-collection-types.md

jkendall327 opened the pull request:
## Summary
Copy-edit for readability and flow by introducing more paragraph breaks, bulletpoints, headings and simplified language. I've also changed the date as this is a fairly major edit.

Please let me know if I've done anything wrong, I've not worked on the .NET docs before!

Fixes #24577

jkendall327 committed:
Update commonly-used-collection-types.md

Copy-edit document for readability and flow by introducing more paragraph breaks, bulletpoints, headings and simplified language.

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/docs?pullRequest=24614) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/docs?pullRequest=24614) before we can accept your contribution.<br/><br/>:x: jkendall327 [sign now](https://cla.dotnetfoundation.org/dotnet/docs?pullRequest=24614)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/docs?pullRequest=24614) it.</sub>

gewarren commented on docs/standard/collections/commonly-used-collection-types.md:
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
- Collections can vary, depending on how the elements are stored, how they are sorted, how searches are performed, and how comparisons are made. The <xref:System.Collections.Queue> class and the <xref:System.Collections.Generic.Queue%601> generic class provide first-in-first-out lists, while the <xref:System.Collections.Stack> class and the <xref:System.Collections.Generic.Stack%601> generic class ...
+The LINQ to Objects feature provides a common pattern for accessing the in-memory objects of any type that implements <xref:System.Collections.IEnumerable> or <xref:System.Collections.Generic.IEnumerable%601>. LINQ queries have several benefits over standard constructs like `foreach` loops:
+
+- They are concise and easier to understand
```
Please add periods to the bullet points.

gewarren commented on docs/standard/collections/commonly-used-collection-types.md:
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
+The LINQ to Objects feature provides a common pattern for accessing the in-memory objects of any type that implements <xref:System.Collections.IEnumerable> or <xref:System.Collections.Generic.IEnumerable%601>. LINQ queries have several benefits over standard constructs like `foreach` loops:
+
+- They are concise and easier to understand
+- They can filter, order and group data
```
```suggestion
- They can filter, order, and group data.
```

gewarren commented on docs/standard/collections/commonly-used-collection-types.md:
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
+## Using LINQ with collection types
   
- Collections can vary, depending on how the elements are stored, how they are sorted, how searches are performed, and how comparisons are made. The <xref:System.Collections.Queue> class and the <xref:System.Collections.Generic.Queue%601> generic class provide first-in-first-out lists, while the <xref:System.Collections.Stack> class and the <xref:System.Collections.Generic.Stack%601> generic class ...
+The LINQ to Objects feature provides a common pattern for accessing the in-memory objects of any type that implements <xref:System.Collections.IEnumerable> or <xref:System.Collections.Generic.IEnumerable%601>. LINQ queries have several benefits over standard constructs like `foreach` loops:
```
```suggestion
The LINQ to Objects feature provides a common pattern for accessing in-memory objects of any type that implements <xref:System.Collections.IEnumerable> or <xref:System.Collections.Generic.IEnumerable%601>. LINQ queries have several benefits over standard constructs like `foreach` loops:
```

gewarren commented on docs/standard/collections/commonly-used-collection-types.md:
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
+
+When you need efficient multi-threaded collection access, use the generic collections in the <xref:System.Collections.Concurrent> namespace.
+
+## Creating generic collections
```
```suggestion
## Strong typing
```

gewarren commented on docs/standard/collections/commonly-used-collection-types.md:
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
+When you need efficient multi-threaded collection access, use the generic collections in the <xref:System.Collections.Concurrent> namespace.
+
+## Creating generic collections
+Generic collections are the best solution to strong typing. A `List<xref:System.Int32>`, for instance, will present a compile-time error if you add any other type to it.
```
```suggestion
Generic collections are the best solution to strong typing. For example, adding any type other than an <xref:System.Int32> to a `List<Int32>` collection causes a compile-time error.
```

gewarren commented on docs/standard/collections/commonly-used-collection-types.md:
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
+## Creating generic collections
+Generic collections are the best solution to strong typing. A `List<xref:System.Int32>`, for instance, will present a compile-time error if you add any other type to it.
+ 
+If your language does not support generics, the <xref:System.Collections> namespace includes abstract base classes that you can extend to create strongly-typed collections: 
```
```suggestion
However, if your language does not support generics, the <xref:System.Collections> namespace includes abstract base classes that you can extend to create collection classes that are strongly typed. These base classes include:
```

gewarren commented on docs/standard/collections/commonly-used-collection-types.md:
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
+## How collections vary
   
- Generic collections are the best solution to strong typing. However, if your language does not support generics, the <xref:System.Collections> namespace includes base collections, such as <xref:System.Collections.CollectionBase>, <xref:System.Collections.ReadOnlyCollectionBase>, and <xref:System.Collections.DictionaryBase>, which are abstract base classes that can be extended to create collectio ...
+Collections vary depending on how they store elements, how they perform searches, how they compare elements, and how they are sorted. 
```
```suggestion
Collections vary in how they store, sort, and compare elements, and how they perform searches. 
```

gewarren commented on docs/standard/collections/commonly-used-collection-types.md:
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
+### Sorting
+You cannot sort the <xref:System.Collections.Hashtable> or the <xref:System.Collections.Generic.Dictionary%602> classes. The <xref:System.Collections.SortedList> class and the generic <xref:System.Collections.Generic.SortedList%602> provide sorted versions of these collections.
+
+The <xref:System.Collections.Queue> and <xref:System.Collections.Generic.Queue%601> classes provide first-in-first-out lists. The <xref:System.Collections.Stack> and <xref:System.Collections.Generic.Stack%601> classes provide last-in-first-out lists. 
```
This doesn't have anything to do with sorting.

gewarren commented on docs/standard/collections/commonly-used-collection-types.md:
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
+All collections use zero-based indexes except <xref:System.Array>, which allows arrays that are not zero-based.  
+
+### Sorting
+You cannot sort the <xref:System.Collections.Hashtable> or the <xref:System.Collections.Generic.Dictionary%602> classes. The <xref:System.Collections.SortedList> class and the generic <xref:System.Collections.Generic.SortedList%602> provide sorted versions of these collections.
```
I don't think that's true. You can call OrderBy to sort dictionaries and hash tables. I would just keep the original version of this sentence.

gewarren commented on docs/standard/collections/commonly-used-collection-types.md:
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
 ---
 # Commonly Used Collection Types
```
```suggestion
# Commonly used collection types
```

gewarren commented on docs/standard/collections/commonly-used-collection-types.md:
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
-Collection types are the common variations of data collections, such as hash tables, queues, stacks, bags, dictionaries, and lists.  
-  
- Collections are based on the <xref:System.Collections.ICollection> interface, the <xref:System.Collections.IList> interface, the <xref:System.Collections.IDictionary> interface, or their generic counterparts. The <xref:System.Collections.IList> interface and the <xref:System.Collections.IDictionary> interface are both derived from the <xref:System.Collections.ICollection> interface; therefore, a ...
+Collection types represent ways to collect data, such as hash tables, queues, stacks, bags, dictionaries, and lists.
```
```suggestion
Collection types represent different ways to collect data, such as hash tables, queues, stacks, bags, dictionaries, and lists.
```

gewarren commented on docs/standard/collections/commonly-used-collection-types.md:
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
- Collections are based on the <xref:System.Collections.ICollection> interface, the <xref:System.Collections.IList> interface, the <xref:System.Collections.IDictionary> interface, or their generic counterparts. The <xref:System.Collections.IList> interface and the <xref:System.Collections.IDictionary> interface are both derived from the <xref:System.Collections.ICollection> interface; therefore, a ...
+Collection types represent ways to collect data, such as hash tables, queues, stacks, bags, dictionaries, and lists.
+
+All collections are based on the <xref:System.Collections.ICollection> interface. <xref:System.Collections.IList>, <xref:System.Collections.IDictionary> and their generic counterparts both derive from <xref:System.Collections.ICollection>. 
```
```suggestion
All collections are based on the <xref:System.Collections.ICollection> or <xref:System.Collections.Generic.ICollection%601> interface, either directly or indirectly. <xref:System.Collections.IList> and <xref:System.Collections.IDictionary> and their generic counterparts all derive from these two interfaces. 
```

gewarren commented on docs/standard/collections/commonly-used-collection-types.md:
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
+
+All collections are based on the <xref:System.Collections.ICollection> interface. <xref:System.Collections.IList>, <xref:System.Collections.IDictionary> and their generic counterparts both derive from <xref:System.Collections.ICollection>. 
+
+In collections based on <xref:System.Collections.IList>, or directly on <xref:System.Collections.ICollection>, every element contains only a value. Examples include:
```
```suggestion
In collections based on <xref:System.Collections.IList> or directly on <xref:System.Collections.ICollection>, every element contains only a value. These types include:
```

gewarren commented on docs/standard/collections/commonly-used-collection-types.md:
```diff
@@ -15,15 +15,68 @@ ms.assetid: f5d4c6a4-0d7b-4944-a9fb-3b12d9ebfd55
+- <xref:System.Collections.Concurrent.ConcurrentStack%601>
+- <xref:System.Collections.Generic.LinkedList%601>
+
+In collections based on the <xref:System.Collections.IDictionary> interface, every element contains both a key and a value. Examples include:
```
```suggestion
In collections based on the <xref:System.Collections.IDictionary> interface, every element contains both a key and a value. These types include:
```

gewarren reviewed (changes requested):
@jkendall327 thanks for your contribution to improving docs. I've left some requested changes.

jkendall327 committed:
Apply suggestions from code review

Apply typographical suggestions

Co-authored-by: Genevieve Warren <24882762+gewarren@users.noreply.github.com>

jkendall327 committed:
Add suggested changes

- Add periods to bulletpoints
- Remove sorting header
- Revert to old phrasing for sorted/unsorted collections

jkendall327 commented:
@gewarren Thanks for the comments, I've just incorporated them all. Let me know if you think any further changes would help.

gewarren committed:
A few more tweaks

gewarren commented:
I made a few more changes directly, hope you don't mind. Looks like there are still some trailing spaces to clean up.

For the periods on the bullet points, that's only if the bullet points are more than a few words. If it's just one word, don't add a period.

jkendall327 committed:
Remove trailing spaces

jkendall327 commented:
No worries, I appreciate the improvements. I've just removed the trailing spaces caught by the linter. Thanks!

gewarren commented:
This looks good to me. @IEvangelist do you want to sign off on this? [Preview link](https://review.docs.microsoft.com/en-us/dotnet/standard/collections/commonly-used-collection-types?branch=pr-en-us-24614).

IEvangelist reviewed (approved):
Looks good to me, we'll :shipit: - thank you 🙏
